### PR TITLE
improve(dependencies): drop truffle from repo and remove truffle build

### PR DESCRIPTION
**Pull request by mrice32**
**Motivation**

We would like to no longer depend on truffle.

**Summary**

This drops the truffle dependency from the repo, removes the truffle build, and drops anything else that was only used by truffle or required truffle.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [ ]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [x]  All existing tests pass
- [ ]  Untested

**Issue(s)**

N/A

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -15,7 +15,7 @@
     "load-addresses": "yarn workspace @uma/core load-addresses",
     "create-release": "lerna version --no-git-tag-version --conventional-commits --no-push",
     "publish-release": "./scripts/run_release.sh",
-    "qbuild": "yarn workspace @uma/common build && yarn workspace @uma/core build && yarn lerna run build --stream --ignore '*/*dapp*' --ignore '@uma/common' --ignore '@uma/core'",
+    "qbuild": "yarn lerna run build --stream",
```
Comment (mrice32) [inline package.json:18]: We can now simplify the build command because there are no circular deps!

File: packages/core/index.ts [hunk 2]
```diff
@@ -1,158 +1 @@
...
-  return truffleContract.abi;
-};
-
-const globalArtifactsExists = Boolean(((global as unknown) as { artifacts: any }).artifacts);
-const getAbi = globalArtifactsExists ? getAbiTest : getAbiProd;
-const getAddress = globalArtifactsExists ? getAddressTest : getAddressProd;
-const getTruffleContract = globalArtifactsExists ? getTruffleContractTest : getTruffleContractProd;
-
-export { getAbi, getAddress, getTruffleContract, findContractVersion };
+export { findContractVersion } from "./src/FindContractVersion";
```
Comment (mrice32) [inline packages/core/index.ts:1]: This is the only function we're still using in core. Not too concerned about relying on it. Maybe down the road, we can move it out.

File: packages/core/migrations/10_deploy_designated_voting_factory.js [hunk 3]
```diff
@@ -1,11 +0,0 @@
-const DesignatedVotingFactory = artifacts.require("DesignatedVotingFactory");
```
Comment (mrice32) [inline packages/core/migrations/10_deploy_designated_voting_factory.js:1]: All migrations were removed.

File: packages/core/package.json [hunk 4]
```diff
@@ -4,24 +4,17 @@
   "description": "UMA smart contracts and unit tests",
   "dependencies": {
     "@openzeppelin/contracts": "4.1.0",
-    "@truffle/contract": "^4.3.15",
     "@uma/common": "^2.7.0",
-    "@uma/core-1-1-0": "npm:@uma/core@1.1.0",
```
Comment (mrice32) [inline packages/core/package.json:9]: No longer need older versions of core to support the index.ts file functions.

File: packages/core/package.json [hunk 5]
```diff
@@ -4,24 +4,17 @@
...
     "@uniswap/v2-periphery": "1.1.0-beta.0",
     "@uniswap/v3-core": "^1.0.0-rc.2",
-    "@uniswap/v3-periphery": "^1.0.0-beta.23",
-    "arb-ts": "^0.0.42",
-    "decimal.js": "^10.2.1",
-    "ethereum-waffle": "^3.4.0"
+    "@uniswap/v3-periphery": "^1.0.0-beta.23"
   },
   "devDependencies": {
+    "arb-ts": "^0.0.42",
```
Comment (mrice32) [inline packages/core/package.json:15]: I don't think these are needed at runtime (i.e. in contracts), so they were moved down to dev dependencies.

File: packages/core/package.json [hunk 6]
```diff
@@ -70,30 +61,27 @@
...
+    "test": "yarn hardhat-test",
     "test-e2e": "mocha -test ./.mocharc.js --e2e",
-    "truffle-test": "truffle test ./truffle-test/scripts/*",
     "hardhat-test": "hardhat test --network hardhat",
     "mocha-test": "mocha ./mocha-test --recursive",
-    "load-addresses": "yarn run apply-registry && yarn hardhat load-addresses",
-    "clean": "rm -rf build && rm -rf contract-types && rm -rf artifacts && rm -rf deployments && rm -rf dist",
-    "build-contracts": "yarn truffle compile && yarn hardhat compile && yarn hardhat compile --network optimism",
+    "load-addresses": "yarn hardhat load-addresses",
+    "clean": "rm -rf contract-types && rm -rf artifacts && rm -rf deployments && rm -rf dist",
```
Comment (mrice32) [inline packages/core/package.json:79]: Most of this is just cleaning up old truffle commands.

File: packages/trader/src/index.ts [hunk 7]
```diff
@@ -4,7 +4,7 @@ import retry from "async-retry";
 import { config } from "dotenv";
 
 import { getWeb3 } from "@uma/common";
-import { getAbi, getAddress } from "@uma/core";
+import { getAbi, getAddress } from "@uma/contracts-node";
```
Comment (mrice32) [inline packages/trader/src/index.ts:7]: This is a drop-in replacement almost everywhere.

File: package.json (hunk 1 above)
Comment (nicholaspai) [reply package.json:18]: v exciting

File: packages/common/src/ProviderUtils.ts [hunk 8]
```diff
@@ -92,10 +92,13 @@ function getDefaultKeyType(): typeof KEY_TYPES[number] {
   return "none";
 }
 
+export function getMnemonic(): string {
```
Comment (nicholaspai) [inline packages/common/src/ProviderUtils.ts:95]: +1 useful function

File: packages/api/package.json [hunk 9]
```diff
@@ -5,6 +5,7 @@
   "author": "David Adams",
   "license": "AGPL-3.0-or-later",
   "devDependencies": {
+    "@types/cors": "^2.8.12",
```
Comment (chrismaree) [inline packages/api/package.json:8]: why was this needed to be added here?

File: packages/core/README.md [hunk 10]
```diff
@@ -4,155 +4,82 @@ This package contains contract artifacts for all UMA smart contracts.
...
 yarn add @uma/core
 ```
 
-## Importing artifacts directly
+## Deploying the contracts
 
-The easiest way to import contracts is to do so by importing the file from
-`@uma/core/build/contracts/ContractName.json`. This is especially useful in certain contexts where imports cannot be
-dynamically loaded (generally non-nodejs contexts):
+To deploy the contracts, we use the `hardhat-deploy` package. This may be failiar to users of `hardhat`. The deployment
```
Comment (nicholaspai) [inline packages/core/README.md:19]:
```suggestion
To deploy the contracts, we use the `hardhat-deploy` package. This may be familiar to users of `hardhat`. The deployment
```

File: packages/core/package.json (hunk 4 above)
Comment (nicholaspai) [reply packages/core/package.json:9]: !!

File: packages/core/package.json (hunk 5 above)
Comment (nicholaspai) [reply packages/core/package.json:15]: Agreed

File: packages/common/src/SolcoverConfig.ts [hunk 11]
```diff
@@ -1,10 +0,0 @@
-export const SolcoverConfig = {
```
Comment (chrismaree) [inline packages/common/src/SolcoverConfig.ts:1]: this is not truffle specific. it actually looks like the last time our coverage ran was feb of this year: https://coveralls.io/github/UMAprotocol/protocol?branch=master

image


we should consider adding a task to get this back up and running.

File: packages/core/README.md (hunk 10 above)
Comment (mrice32) [reply packages/core/README.md:19]: Lol failiar

File: packages/common/src/TruffleConfig.ts [hunk 12]
```diff
@@ -1,242 +0,0 @@
-/**
```
Comment (chrismaree) [inline packages/common/src/TruffleConfig.ts:1]: 👋

**Review by nicholaspai — APPROVED** [review]
VERY satisfying PR this is amazing. I'm so excited, LGTM

File: packages/core/index.ts (hunk 2 above)
Comment (chrismaree) [reply packages/core/index.ts:1]: ye, could easily move it common

**Review by chrismaree — APPROVED** [review]
this is amazing! the culmination of many many hours of painful work all come together in this RR. really happy to see all these migrations and truffle dependent things being removed.

In other related news, I've spent a fair amount of time getting familiar with the hardhat console and I'm happy to report that for many actions it's actually much better than the truffle console!

File: packages/api/package.json (hunk 9 above)
Comment (mrice32) [reply packages/api/package.json:8]: Good question. The api build began failing because it couldn't resolve the types for the cors package. My (and @daywiss's) assumption is that this was previously getting this dep via some nested package that was installed under truffle, so once it was removed this began failing.

Anyway, the package uses the cors types, so it should be including them as a dep. It was a little odd to see it sporadically fail like this though.
